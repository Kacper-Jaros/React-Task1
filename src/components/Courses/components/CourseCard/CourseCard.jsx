import PropTypes from 'prop-types';
import styled from 'styled-components';
import { useNavigate } from 'react-router';
import { useDispatch } from 'react-redux';

import { Button } from 'common/Button/Button';
import { COURSE_CARD_FIELD, BUTTON_TEXT } from 'constans';

import { validateAuthors } from 'helpers/authors';
import { timeConvert } from 'helpers/pipeDuration';
import { deleteCourse } from 'store/courses/thunk';

export const CourseCard = ({
	title,
	duration,
	destription,
	authors,
	created,
	id,
	userRole,
}) => {
	const names = validateAuthors(authors);
	const time = timeConvert(duration);
	const navigate = useNavigate();
	const dispatch = useDispatch();

	const handleOnClick = () => {
		navigate(`/courses/${id}`);
	};

	const handleDeleteCourse = () => {
		dispatch(deleteCourse(id));
	};

	const handleUpdateCourse = () => {
		navigate(`/courses/update/${id}`);
	};
	return (
		<CourseCardContainer>
			<TitelAndDesscription>
				<StyledTitle>{title}</StyledTitle>
				<StyledDescription>{destription}</StyledDescription>
			</TitelAndDesscription>
			<AuthorsAndButton>
				<div>
					<TextField>{COURSE_CARD_FIELD.AUTHORS}: </TextField>
					<text>{names}</text>
				</div>
				<div>
					<TextField>{COURSE_CARD_FIELD.DURATION}: </TextField>
					<text>{time}</text>
				</div>
				<div>
					<TextField>{COURSE_CARD_FIELD.CREATED}: </TextField>
					<text>{created}</text>
				</div>
				<div>
					<StyledButton
						buttonText={BUTTON_TEXT.SHOW_COURSE}
						handleClick={handleOnClick}
					/>
					{userRole.toLowerCase() === 'admin' ? (
						<>
							<StyledButton buttonText={'U'} handleClick={handleUpdateCourse} />
							<StyledButton buttonText={'D'} handleClick={handleDeleteCourse} />
						</>
					) : null}
				</div>
			</AuthorsAndButton>
		</CourseCardContainer>
	);
};

CourseCard.propTypes = {
	authors: PropTypes.array,
	created: PropTypes.string,
	destription: PropTypes.string,
	duration: PropTypes.any,
	id: PropTypes.string,
	title: PropTypes.string,
};

const CourseCardContainer = styled.div`
	display: flex;
	border: 1px solid green;
	padding: 20px;
	margin: 20px;
`;

const TitelAndDesscription = styled.div`
	width: 70%;
`;
const AuthorsAndButton = styled.div`
	display: flex;
	flex-direction: column;
	margin: auto;
`;

const StyledTitle = styled.div`
	font-weight: bold;
	font-size: x-large;
`;

const StyledDescription = styled.div`
	font-size: medium;
`;

const TextField = styled.text`
	font-weight: bold;
`;

const StyledButton = styled(Button)`
	align-self: center;
	margin-top: 20px;
	margin-left: 10px;
`;
