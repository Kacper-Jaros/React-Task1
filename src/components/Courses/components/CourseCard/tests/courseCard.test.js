import 'jest';
import { CourseCard } from '../CourseCard';
import { Router } from 'react-router';
import { render, screen } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { Provider } from 'react-redux';

const mockedState = {
	userReducer: {
		isAUth: true,
		name: 'Test Name',
	},
	coursesReducer: [],
	authorsReducer: [
		{
			id: '27cc3006-e93a-4748-8ca8-73d06aa93b6d',
			name: 'Vasiliy Dobkin',
		},
	],
};

const mockedStore = {
	getState: () => mockedState,
	subscribe: jest.fn(),
	dispatch: jest.fn(),
};

test(`CourseCard should display title.`, () => {
	const history = createMemoryHistory();
	history.push('/courses');

	render(
		<Router location={history.location} navigator={history}>
			<Provider store={mockedStore}>
				<CourseCard
					title={'title'}
					authors={mockedState.authorsReducer}
					userRole={'user'}
				/>
			</Provider>
		</Router>
	);

	expect(screen.queryByText('title')).toBeInTheDocument();
});

test(`CourseCard should display description.`, () => {
	const history = createMemoryHistory();
	history.push('/courses');

	render(
		<Router location={history.location} navigator={history}>
			<Provider store={mockedStore}>
				<CourseCard
					title={'title'}
					destription={'description'}
					authors={mockedState.authorsReducer}
					userRole={'user'}
				/>
			</Provider>
		</Router>
	);

	expect(screen.queryByText('description')).toBeInTheDocument();
});

test(`CourseCard should display duration in the correct format.`, () => {
	const history = createMemoryHistory();
	history.push('/courses');

	render(
		<Router location={history.location} navigator={history}>
			<Provider store={mockedStore}>
				<CourseCard
					title={'title'}
					duration={70}
					destription={'description'}
					authors={mockedState.authorsReducer}
					userRole={'user'}
				/>
			</Provider>
		</Router>
	);

	expect(screen.queryByText('01:10 hours')).toBeInTheDocument();
});

test(`CourseCard should display authors list.`, () => {
	const history = createMemoryHistory();
	history.push('/courses');

	render(
		<Router location={history.location} navigator={history}>
			<Provider store={mockedStore}>
				<CourseCard
					title={'title'}
					authors={mockedState.authorsReducer}
					userRole={'user'}
				/>
			</Provider>
		</Router>
	);

	expect(screen.queryByText('Vasiliy Dobkin')).toBeInTheDocument();
});

test(`CourseCard should display created date in the correct format.`, () => {
	const history = createMemoryHistory();
	history.push('/courses');

	render(
		<Router location={history.location} navigator={history}>
			<Provider store={mockedStore}>
				<CourseCard
					title={'title'}
					duration={70}
					destription={'description'}
					authors={mockedState.authorsReducer}
					userRole={'user'}
					created={'01.02.2018'}
				/>
			</Provider>
		</Router>
	);

	expect(screen.queryByText('01.02.2018')).toBeInTheDocument();
});
