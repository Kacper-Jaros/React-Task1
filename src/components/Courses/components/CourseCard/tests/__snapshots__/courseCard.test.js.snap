// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseCard should display duration in the correct format. 1`] = `null`;
